Create ACL_2_8 python3 test module

j-ororke opened the pull request:
Creating TC_ACL_2_8 python3 test module: 
- Adds TC_ACL_2_8 python3 test module 
- This used base branch of Amine's PR [38263](https://github.com/project-chip/connectedhomeip/pull/38263) this is a follow-up PR to that PR.

#### Testing
WSL - Windows Subsystem Linux

Alami-Amine committed:
Testing removing one of the template functions

Alami-Amine committed:
WriteClient: Encoding as many items as possible into a single list as part of ReplaceAll item operation

Alami-Amine committed:
fixing failures

Alami-Amine committed:
Fix to Error Handling in User-Label Cluster: return RESOURCE_EXHAUSTED when the number of items in list exceeds kMaxUserLabelListLength

Alami-Amine committed:
chip-repl: support for preencoded attributes

Alami-Amine committed:
Fixing ACL 2.3 and 2.5

Alami-Amine committed:
Validating ACLs Upfront: This is done to make sure we reject the whole list if any of the entries within it are non conforming

Alami-Amine committed:
more yaml test changes

restyled-commits committed:
Restyled by prettier-yaml

Alami-Amine committed:
Fix Unit Testing Cluster (Test Cluster Server) to process ReplaceAll operations

Alami-Amine committed:
fix PutPreencodedAttribute

Alami-Amine committed:
Cleaning code

Alami-Amine committed:
More Code Clean Up

Alami-Amine committed:
Unit Test Cleanup

Alami-Amine committed:
more cleanup

Alami-Amine committed:
Adding testcases to TestAccessControlCluster

Alami-Amine committed:
Fixing comments

Alami-Amine committed:
Making IsValid method in AccessControl Static

Alami-Amine committed:
Adding comments

Alami-Amine committed:
Make IsValid a member function of AccessControl::Entry

j-ororke committed:
Creating TC_ACL_2_8 python3 test module:
- Adds TC_ACL_2_8 python3 test module
- This uses base branch of Amine's PR #37755 to create this test module: This is a follow-up PR to that PR from Amine, once Amine's PR is merged this one can come out of WIP state.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_ACL_2_8 test module:
- Resolving linting errors

j-ororke committed:
Updating TC_ACL_2_8 test module:
- Resolving more linting errors

restyled-commits committed:
Restyled by isort

Alami-Amine commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -0,0 +1,312 @@
+    '17D00000F1FF01003D48656C6C6F20576F726C642E205468697320697320612073696E676C6520656C656D656E74206C6976696E6720617320612063686172737472696E670018')
+
+
+class TC_ACL_2_8(MatterBaseTest):
```
I think you converted ACL_2_7 here instead of ACL_2_8 :smile:

j-ororke replied:
Sorry, I think you are correct lol I am not sure how the code got misplaced in this PR, but indeed it did, and has now been updated. Hopefully, this one contains the correct test steps and test code in it this time lol

Alami-Amine replied:
looks good now :smile:

j-ororke committed:
Updating TC_ACL_2_8 test module:
- Adding correct code for TC_ACL_2_8 test module, in place of TC_ACL_2_7 code.

j-ororke committed:
Updating TC_ACL_2_8 python3 test module:
- Removing some unneeded comments in the code

j-ororke committed:
Updating TC_ACL_2_8 python3 test module:
- refining _verify_acl_event() private method

j-ororke committed:
Updating TC_ACL_2_8 test module:
- Used fixed patch file to apply restylizer am patch to code

j-ororke committed:
resolving linting errors

j-ororke committed:
Running isort and autopep8 on the code to resolve the stylizer issues since the generated patch is having issues

Alami-Amine commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -0,0 +1,324 @@
+            TestStep(8, "TH2 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 1 element, and MUST NOT contain an element with FabricIndex F1"),
+            TestStep(9, "TH1 reads DUT Endpoint 0 AccessControl cluster AccessControlEntryChanged event",
+                     "Result is SUCCESS, value is list of AccessControlEntryChanged containing 3 elements, and MUST NOT contain any element with FabricIndex F2"),
```
```suggestion
                     "Result is SUCCESS, value is list of AccessControlEntryChanged containing 2 elements, and MUST NOT contain any element with FabricIndex F2"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -0,0 +1,324 @@
+            admin_passcode_id,
+            change_type,
+            subjects,
+            targets,
```
you are not using `targets`, remove it probably?

j-ororke replied:
Thank you, removed targets arg.

j-ororke committed:
Apply suggestions from code review

Changing test step as per Amine's comments

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Updating TC_ACL_2_8 python3 test module:
- Resolving formating issues
- Changed to using self.read_single_attribute_check_success() from matter_testing support in place of ReadAttribute()
- Removed targets arg from _verify_acl_event pvt function.
- Removed exceptions as errors caught by asserts.

j-ororke committed:
Updating TC_ACL_2_8 python3 test module:
- Removing additional try and except block

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_ACL_2_8 python3 test module:
- Changing some self.print_step() to logging.info()

j-ororke committed:
Updating TC_ACL_2_8 python3 test module: - Removing try and except blocks from the code

j-ororke committed:
Syncing with upstream master to resolve file conflicts

Alami-Amine reviewed (changes requested):
Should we remove the YAML as well?

j-ororke committed:
Removing Test_TC_ACL_2_8 yaml script

cecille commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -0,0 +1,287 @@
+        data = event.Data
+
+        # Verify admin IDs
+        if admin_node_id is None:
```
why not just pass in NullValue from the caller?

j-ororke replied:
That is a really good question, I honestly don't know why I didn't do that here. 
I have updated the test to do that now though, it removes those nasty conditional checks in the function, and cleans the code up a bit, also makes it cleaner at the call site as well!

cecille commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -0,0 +1,287 @@
+
+        self.step(7)
+        # TH1 reads ACL attribute
+        ac_cluster = Clusters.AccessControl
```
Same comment as the extensions test - can you create a follow up issue to adjust the test plan to test both fabric filtering and fabric sensitive field omissions? Don't block this PR on implementing - let's do this after TE2. Just link the issue.

j-ororke replied:
Created following follow-up task to enhance tests to cover testing fabric filtered and fabric sensitive scenarios:
- https://github.com/project-chip/matter-test-scripts/issues/549

j-ororke committed:
Updated to using Nullvalue from caller in _verify_acl_event() making the code cleaner, thank you Cecille!

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_8_test

Alami-Amine commented:
- Backward Compatibility: Passed Tests when run against `chip-all-clusters-app` from Matter 1.0
